// @qawolf/ci-sdk/dist/esm/lib/sdk/domain/legacy_previewTesting/index.js
import { attemptNotifyDeploy } from "../attempt-deploy";
import { previewDeploymentType } from "./constants";
import { createEnvironmentAction as createEnvironment } from "./createEnvironmentAction";
import { deleteEnvironmentAction } from "./deleteEnvironmentAction";
export async function testPreview(deps, apiConfig, config) {
    await createEnvironment(deps, apiConfig, {
        baseEnvironmentId: config.baseEnvironmentId,
        branch: config.branch,
        deploymentUrl: config.deploymentUrl,
        headRepoFullName: config.repoFullName,
        pr: config.pr,
        qaWolfTeamId: config.qaWolfTeamId,
        variables: config.variables,
    });
    return attemptNotifyDeploy(deps, apiConfig, {
        branch: config.branch,
        commitUrl: config.commitUrl,
        deploymentType: previewDeploymentType,
        deploymentUrl: config.deploymentUrl,
        sha: config.sha,
        variables: config.variables,
    });
}
export async function removeEnvironment(deps, apiConfig, config) {
    await deleteEnvironmentAction(deps, apiConfig, config);
}
