// @qawolf/ci-sdk/dist/cjs/lib/sdk/domain/legacy_previewTesting/findRepositoryIdByName.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.findRepositoryIdByName = findRepositoryIdByName;
const constants_1 = require("./constants");
async function findRepositoryIdByName(deps, apiConfig, { headRepoFullName, }) {
    const response = await deps.fetch(constants_1.qawolfGraphQLEndpoint, {
        body: JSON.stringify({
            query: `
        query codeHostingServiceRepositories($where: CodeHostingServiceRepositoryWhereInput) {
          codeHostingServiceRepositories(where: $where) {
            id
          }
        }
      `,
            variables: {
                where: {
                    externalFullName: {
                        equals: headRepoFullName,
                    },
                },
            },
        }),
        headers: {
            Authorization: `Bearer ${apiConfig.apiKey}`,
            "Content-Type": "application/json",
        },
        method: "post",
    });
    const responseJson = (await response.json());
    deps.log.info(`Repository response: ${JSON.stringify(responseJson)}`);
    const repositories = responseJson.data.codeHostingServiceRepositories;
    if (!repositories[0]) {
        return;
    }
    return repositories[0].id;
}
