// @qawolf/ci-sdk/dist/cjs/lib/sdk/domain/legacy_previewTesting/getEnvironmentVariablesFromEnvironment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getEnvironmentVariablesFromEnvironment = getEnvironmentVariablesFromEnvironment;
const constants_1 = require("./constants");
async function getEnvironmentVariablesFromEnvironment(deps, apiConfig, { environmentId, }) {
    const retrievalResponse = await deps.fetch(constants_1.qawolfGraphQLEndpoint, {
        body: JSON.stringify({
            query: `
        query EnvironmentVariables($where: EnvironmentWhereUniqueInput!) {
          environment(where: $where) {
            id
            variablesJSON
          }
        }
      `,
            variables: {
                where: {
                    id: environmentId,
                },
            },
        }),
        headers: {
            Authorization: `Bearer ${apiConfig.apiKey}`,
            "Content-Type": "application/json",
        },
        method: "post",
    });
    const retrievalResponseJson = (await retrievalResponse.json());
    if (!retrievalResponseJson.data.environment) {
        throw Error(`Environment not found with ID: ${environmentId}. Please check the environment ID is correct.`);
    }
    return JSON.parse(retrievalResponseJson.data.environment.variablesJSON);
}
