// @qawolf/ci-sdk/dist/cjs/lib/sdk/domain/legacy_previewTesting/getTagsFromEnvironment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTagsFromGenericTriggerInEnvironment = getTagsFromGenericTriggerInEnvironment;
const constants_1 = require("./constants");
async function getTagsFromGenericTriggerInEnvironment(deps, apiConfig, { environmentId, }) {
    const retrievalResponse = await deps.fetch(constants_1.qawolfGraphQLEndpoint, {
        body: JSON.stringify({
            query: `
        query GenericTriggerTagsFromEnvironment($where: EnvironmentWhereUniqueInput!) {
          environment(where: $where) {
            id
            triggers(where: {deployment_provider: {equals: "generic"}}) {
              id
              tags {
                id
                name
              }
            }
          }
        }
      `,
            variables: {
                where: {
                    id: environmentId,
                },
            },
        }),
        headers: {
            Authorization: `Bearer ${apiConfig.apiKey}`,
            "Content-Type": "application/json",
        },
        method: "post",
    });
    const retrievalResponseJson = (await retrievalResponse.json());
    if (!retrievalResponseJson.data.environment) {
        throw Error(`Environment not found with ID: ${environmentId}. Please check the environment ID is correct.`);
    }
    return retrievalResponseJson.data.environment.triggers[0]?.tags?.map((tag) => tag);
}
