// @qawolf/ci-sdk/dist/cjs/lib/sdk/domain/legacy_previewTesting/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.testPreview = testPreview;
exports.removeEnvironment = removeEnvironment;
const attempt_deploy_1 = require("../attempt-deploy");
const constants_1 = require("./constants");
const createEnvironmentAction_1 = require("./createEnvironmentAction");
const deleteEnvironmentAction_1 = require("./deleteEnvironmentAction");
async function testPreview(deps, apiConfig, config) {
    await (0, createEnvironmentAction_1.createEnvironmentAction)(deps, apiConfig, {
        baseEnvironmentId: config.baseEnvironmentId,
        branch: config.branch,
        deploymentUrl: config.deploymentUrl,
        headRepoFullName: config.repoFullName,
        pr: config.pr,
        qaWolfTeamId: config.qaWolfTeamId,
        variables: config.variables,
    });
    return (0, attempt_deploy_1.attemptNotifyDeploy)(deps, apiConfig, {
        branch: config.branch,
        commitUrl: config.commitUrl,
        deploymentType: constants_1.previewDeploymentType,
        deploymentUrl: config.deploymentUrl,
        sha: config.sha,
        variables: config.variables,
    });
}
async function removeEnvironment(deps, apiConfig, config) {
    await (0, deleteEnvironmentAction_1.deleteEnvironmentAction)(deps, apiConfig, config);
}
