// @qawolf/ci-utils/dist/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.jsonEnvironmentsMappingSchema = exports.jsonEnvironmentVariablesSchema = void 0;
const tslib_1 = require("tslib");
const zod_1 = tslib_1.__importDefault(require("zod"));
const environmentVariablesSchema = zod_1.default.record(zod_1.default.string(), zod_1.default.string());
exports.jsonEnvironmentVariablesSchema = zod_1.default
    .string()
    .transform((str, ctx) => {
    try {
        return JSON.parse(str);
    }
    catch (error) {
        ctx.addIssue({
            code: "custom",
            message: "input is not a valid JSON-formatted string",
        });
        return;
    }
})
    .pipe(environmentVariablesSchema);
const environmentsMappingSchema = zod_1.default.array(zod_1.default.object({
    environmentAlias: zod_1.default.string(),
    vcsBranch: zod_1.default.string(),
}));
exports.jsonEnvironmentsMappingSchema = zod_1.default
    .string()
    .transform((str, ctx) => {
    try {
        return JSON.parse(str);
    }
    catch (error) {
        ctx.addIssue({
            code: "custom",
            message: "input is not a valid JSON-formatted string",
        });
        return;
    }
})
    .pipe(environmentsMappingSchema);
