// @qawolf/ci-sdk/dist/esm/lib/api/notify-vcs-branch-build-deployed-mutation.js
import { qawolfGraphql } from "./graphql";
const mutationName = "notifyVCSBranchBuildDeployed";
const mutationGql = `
mutation NotifyVCSBranchBuildDeployed($headVcsCommitId: String!, $headEnvironmentVariablesJson: String!, $baseEnvironmentAlias: String!, $headEnvironmentAlias: String!) {
    ${mutationName}(headVcsCommitId: $headVcsCommitId, headEnvironmentVariablesJson: $headEnvironmentVariablesJson, baseEnvironmentAlias: $baseEnvironmentAlias, headEnvironmentAlias: $headEnvironmentAlias) {
    outcome
    codeHostingServiceInstallationPlatform
    failureCode
    failureDetails
    runId
  }
}`;
export async function callNotifyVCSBranchBuildDeployedMutation(deps, apiConfig, { baseEnvironmentAlias, concurrencyLimit, headEnvironmentAlias, headEnvironmentName, headEnvironmentVariables, headVcsCommitId, headVcsCommitUrl, headVcsBranch, pullOrMergeRequestNumber, }) {
    return qawolfGraphql({
        apiConfig,
        deps,
        name: mutationName,
        query: mutationGql,
        variables: {
            baseEnvironmentAlias,
            concurrencyLimit: concurrencyLimit ?? null,
            headEnvironmentAlias,
            headEnvironmentName,
            headEnvironmentVariablesJson: JSON.stringify(headEnvironmentVariables),
            headVcsBranch: headVcsBranch ?? null,
            headVcsCommitId,
            headVcsCommitUrl: headVcsCommitUrl ?? null,
            pullOrMergeRequestNumber: pullOrMergeRequestNumber ?? null,
        },
        // The type coercion highlights that we don't have runtime schema validation
    });
}
