// @qawolf/ci-sdk/dist/esm/lib/sdk/domain/vcsBranchTesting/notify-vcs-branch-merge-completed.js
import { callNotifyVCSBranchMergeCompletedMutation } from "../../../api/notify-vcs-branch-merge-completed-mutation";
import { assertType } from "../../../utils";
import { domainFailureToAbortResult } from "./lib/domain-failure";
import { graphQLErrorToAbortResult } from "./lib/graphql-error";
import { retryWithExponentialBackoff } from "./lib/retry";
async function runNotifyVCSBranchMergeCompletedOnce(deps, apiConfig, input) {
    const log = deps.log;
    const { baseEnvironmentsMapping, baseVcsBranch: vcsBaseBranch } = input;
    const baseEnvironmentAlias = baseEnvironmentsMapping.find((mapping) => mapping.vcsBranch === vcsBaseBranch)?.environmentAlias;
    if (baseEnvironmentAlias === undefined) {
        log.error(`❌ [notifyVCSBranchMergeCompleted] Could not find a base environment for VCS branch '${vcsBaseBranch}'. Make sure you provide ` +
            "a mapping for this base branch in the 'baseEnvironmentsMapping' field.");
        return {
            abortReason: "missing-vcs-branch-to-environment-alias-mapping",
            outcome: "aborted",
        };
    }
    const resp = await callNotifyVCSBranchMergeCompletedMutation(deps, apiConfig, {
        baseEnvironmentAlias,
        headEnvironmentAlias: input.headEnvironmentAlias,
    });
    if (resp.isGqlError) {
        return graphQLErrorToAbortResult({
            graphQLPayload: resp,
            log,
            methodName: "notifyVCSBranchMergeCompleted",
        });
    }
    const result = resp.responseBody;
    if (result.outcome === "success") {
        log.info(`✅ [notifyVCSBranchMergeCompleted] Successfully notified the CI system that the merge was completed.`);
        return {
            outcome: "success",
        };
    }
    assertType(result.outcome);
    return domainFailureToAbortResult({
        log,
        methodName: "notifyVCSBranchMergeCompleted",
        result,
    });
}
export async function notifyVCSBranchMergeCompleted(deps, apiConfig, input) {
    const { maxRetries = 10 } = input;
    return retryWithExponentialBackoff({
        log: deps.log,
        maxRetries,
        methodName: "notifyVCSBranchMergeCompleted",
        retriableAbortReasons: [
            "network-error",
            "server-error",
        ],
        runOnce: () => runNotifyVCSBranchMergeCompletedOnce(deps, apiConfig, input),
    });
}
