// @qawolf/ci-sdk/dist/cjs/lib/sdk/domain/vcsBranchTesting/notify-vcs-branch-merge-completed.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.notifyVCSBranchMergeCompleted = notifyVCSBranchMergeCompleted;
const notify_vcs_branch_merge_completed_mutation_1 = require("../../../api/notify-vcs-branch-merge-completed-mutation");
const utils_1 = require("../../../utils");
const domain_failure_1 = require("./lib/domain-failure");
const graphql_error_1 = require("./lib/graphql-error");
const retry_1 = require("./lib/retry");
async function runNotifyVCSBranchMergeCompletedOnce(deps, apiConfig, input) {
    const log = deps.log;
    const { baseEnvironmentsMapping, baseVcsBranch: vcsBaseBranch } = input;
    const baseEnvironmentAlias = baseEnvironmentsMapping.find((mapping) => mapping.vcsBranch === vcsBaseBranch)?.environmentAlias;
    if (baseEnvironmentAlias === undefined) {
        log.error(`❌ [notifyVCSBranchMergeCompleted] Could not find a base environment for VCS branch '${vcsBaseBranch}'. Make sure you provide ` +
            "a mapping for this base branch in the 'baseEnvironmentsMapping' field.");
        return {
            abortReason: "missing-vcs-branch-to-environment-alias-mapping",
            outcome: "aborted",
        };
    }
    const resp = await (0, notify_vcs_branch_merge_completed_mutation_1.callNotifyVCSBranchMergeCompletedMutation)(deps, apiConfig, {
        baseEnvironmentAlias,
        headEnvironmentAlias: input.headEnvironmentAlias,
    });
    if (resp.isGqlError) {
        return (0, graphql_error_1.graphQLErrorToAbortResult)({
            graphQLPayload: resp,
            log,
            methodName: "notifyVCSBranchMergeCompleted",
        });
    }
    const result = resp.responseBody;
    if (result.outcome === "success") {
        log.info(`✅ [notifyVCSBranchMergeCompleted] Successfully notified the CI system that the merge was completed.`);
        return {
            outcome: "success",
        };
    }
    (0, utils_1.assertType)(result.outcome);
    return (0, domain_failure_1.domainFailureToAbortResult)({
        log,
        methodName: "notifyVCSBranchMergeCompleted",
        result,
    });
}
async function notifyVCSBranchMergeCompleted(deps, apiConfig, input) {
    const { maxRetries = 10 } = input;
    return (0, retry_1.retryWithExponentialBackoff)({
        log: deps.log,
        maxRetries,
        methodName: "notifyVCSBranchMergeCompleted",
        retriableAbortReasons: [
            "network-error",
            "server-error",
        ],
        runOnce: () => runNotifyVCSBranchMergeCompletedOnce(deps, apiConfig, input),
    });
}
