// @qawolf/ci-sdk/dist/cjs/lib/sdk/domain/vcsBranchTesting/notify-vcs-branch-build-deployed.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.notifyVCSBranchBuildDeployed = notifyVCSBranchBuildDeployed;
const notify_vcs_branch_build_deployed_mutation_1 = require("../../../api/notify-vcs-branch-build-deployed-mutation");
const utils_1 = require("../../../utils");
const domain_failure_1 = require("./lib/domain-failure");
const graphql_error_1 = require("./lib/graphql-error");
const retry_1 = require("./lib/retry");
async function runNotifyVCSBranchBuildDeployedOnce(deps, apiConfig, input) {
    const { baseEnvironmentsMapping, baseVcsBranch, headEnvironmentAlias, headEnvironmentName, headEnvironmentVariables, headVcsCommitId, concurrencyLimit, headVcsCommitUrl, headVcsBranch, pullOrMergeRequestNumber, } = input;
    const log = deps.log;
    const baseEnvironmentAlias = baseEnvironmentsMapping.find((mapping) => mapping.vcsBranch === baseVcsBranch)?.environmentAlias;
    if (baseEnvironmentAlias === undefined) {
        log.error(`❌ [notifyVCSBranchBuildDeployed] Could not find a base environment for VCS branch '${baseVcsBranch}'. Make sure you provide ` +
            "a mapping for this base branch in the 'baseEnvironmentsMapping' field.");
        return {
            abortReason: "missing-vcs-branch-to-environment-alias-mapping",
            outcome: "aborted",
        };
    }
    let finalConcurrencyLimit = concurrencyLimit;
    if (typeof concurrencyLimit === "number") {
        if (concurrencyLimit === Infinity) {
            finalConcurrencyLimit = 0;
        }
        else if (isNaN(concurrencyLimit) || !Number.isInteger(concurrencyLimit)) {
            log.error(`❌ [notifyVCSBranchBuildDeployed] Invalid concurrency limit '${concurrencyLimit}'. Must be a positive integer.`);
            return {
                abortReason: "invalid-input",
                outcome: "aborted",
            };
        }
        if (concurrencyLimit < 0) {
            log.error(`❌ [notifyVCSBranchBuildDeployed] Invalid concurrency limit '${concurrencyLimit}'. Must be a positive integer.`);
            return {
                abortReason: "invalid-input",
                outcome: "aborted",
            };
        }
    }
    const resp = await (0, notify_vcs_branch_build_deployed_mutation_1.callNotifyVCSBranchBuildDeployedMutation)(deps, apiConfig, {
        baseEnvironmentAlias,
        concurrencyLimit: finalConcurrencyLimit,
        headEnvironmentAlias,
        headEnvironmentName,
        headEnvironmentVariables,
        headVcsBranch,
        headVcsCommitId,
        headVcsCommitUrl,
        pullOrMergeRequestNumber,
    });
    if (resp.isGqlError) {
        return (0, graphql_error_1.graphQLErrorToAbortResult)({
            graphQLPayload: resp,
            log,
            methodName: "notifyVCSBranchBuildDeployed",
        });
    }
    const result = resp.responseBody;
    if (result.outcome === "success") {
        const codeHostingInfo = result.codeHostingServiceInstallationPlatform === undefined
            ? ""
            : `, and code hosting service integration '${result.codeHostingServiceInstallationPlatform}'.`;
        log.info(`✅ [notifyVCSBranchBuildDeployed] Success. Run was deployed with ID: '${result.runId}'${codeHostingInfo}`);
        return {
            codeHostingServiceInstallationType: result.codeHostingServiceInstallationPlatform,
            outcome: "success",
            runId: result.runId,
        };
    }
    (0, utils_1.assertType)(result.outcome);
    return (0, domain_failure_1.domainFailureToAbortResult)({
        log,
        methodName: "notifyVCSBranchBuildDeployed",
        result,
    });
}
async function notifyVCSBranchBuildDeployed(deps, apiConfig, input) {
    const { maxRetries = 10 } = input;
    return (0, retry_1.retryWithExponentialBackoff)({
        log: deps.log,
        maxRetries,
        methodName: "notifyVCSBranchBuildDeployed",
        retriableAbortReasons: [
            "network-error",
            "server-error",
            "run-creation-failed",
        ],
        runOnce: () => runNotifyVCSBranchBuildDeployedOnce(deps, apiConfig, input),
    });
}
