// @qawolf/ci-sdk/dist/cjs/tests/test-lib.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeVCSBranchTestingSDKForTest = makeVCSBranchTestingSDKForTest;
exports.makeQaWolfSdkForTest = makeQaWolfSdkForTest;
exports.buildResponse = buildResponse;
exports.buildCiGreenlightStatusBase = buildCiGreenlightStatusBase;
const globals_1 = require("@jest/globals");
const vcsBranchTesting_1 = require("../lib/sdk/domain/vcsBranchTesting");
const index_1 = require("../lib/sdk/index");
const localDummyServiceBase = "http://localhost:80/";
const testLogger = {
    error: console.log.bind(console),
    info: console.log.bind(console),
    warn: console.log.bind(console),
};
function makeVCSBranchTestingSDKForTest({ apiKey }) {
    // Can be used for mocking fetch
    const fetch = globals_1.jest.fn(() => Promise.resolve(buildResponse({})));
    return {
        fetch,
        sdk: (0, vcsBranchTesting_1.makeVCSBranchTestingSDK)({
            apiKey,
            serviceBase: localDummyServiceBase,
        }, {
            fetch,
            log: testLogger,
        }),
    };
}
function makeQaWolfSdkForTest({ apiKey }) {
    // Can be used for mocking fetch
    const fetch = globals_1.jest.fn(() => Promise.resolve(buildResponse({})));
    return {
        fetch,
        sdk: (0, index_1.makeQaWolfSdk)({
            apiKey,
            serviceBase: localDummyServiceBase,
        }, {
            fetch,
            log: testLogger,
        }),
    };
}
function buildResponse(response, status = 200) {
    return new Response(JSON.stringify(response), {
        headers: {
            "Content-type": "application/json",
        },
        status,
    });
}
function buildCiGreenlightStatusBase() {
    return {
        relevantRunId: "foo",
        relevantRunUrl: "http://foo",
        relevantRunWithBugsUrl: "http://foo",
        rootRunId: "foo",
        rootRunUrl: "http://foo",
        workflowsDisabledAfterRunCount: 0,
        workflowsInRunCount: 10,
        workflowsUnderInvestigationCount: 0,
    };
}
